pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/ConsistentDataPushUtils.java

```diff
@@ -0,0 +1,285 @@
+          LOGGER.info("Got response {}: {} while sending revert replace segment request for table: {}, uploadURI: {}",
+              response.getStatusCode(), response.getResponse(), rawTableName, entry.getKey());
+        } catch (URISyntaxException | HttpErrorStatusException | IOException e) {
+          LOGGER.error("Exception when sending revert replace segment request to controller: {} for table: {}",
+              entry.getKey(), rawTableName, e);
+        }
+      }
+    }
+  }
+
+  public static TableConfig getTableConfig(SegmentGenerationJobSpec spec) {
+    String rawTableName = spec.getTableSpec().getTableName();
+    TableConfig tableConfig =
+        SegmentGenerationUtils.getTableConfig(spec.getTableSpec().getTableConfigURI(), spec.getAuthToken());
+    Preconditions.checkState(tableConfig != null, "Failed to find table config for table: %s", rawTableName);
+    return tableConfig;
+  }
+
+  public static boolean consistentDataPushEnabled(SegmentGenerationJobSpec spec) {
+    TableConfig tableConfig = getTableConfig(spec);
+    // Enable consistent data push only if "consistentDataPush" is set to true in batch ingestion config and the
+    // table is REFRESH use case.
+    // TODO: Remove the check for REFRESH when we support consistent push for APPEND table
+    return "REFRESH".equalsIgnoreCase(IngestionConfigUtils.getBatchSegmentIngestionType(tableConfig))
+        && IngestionConfigUtils.getBatchSegmentIngestionConsistentDataPushEnabled(tableConfig);
+  }
+
+  /**
+   * Returns a map of controller URI to a list of existing OFFLINE segments.
+   */
+  public static Map<URI, List<String>> getSegmentsToReplace(SegmentGenerationJobSpec spec, String rawTableName) {
+    Map<URI, List<String>> uriToOfflineSegments = new HashMap<>();
+    for (PinotClusterSpec pinotClusterSpec : spec.getPinotClusterSpecs()) {
+      URI controllerURI;
+      List<String> offlineSegments;
+      try {
+        controllerURI = new URI(pinotClusterSpec.getControllerURI());
+        Map<String, List<String>> segments =
+            FILE_UPLOAD_DOWNLOAD_CLIENT.getSegments(controllerURI, rawTableName, TableType.OFFLINE, true);
+        offlineSegments = segments.get(TableType.OFFLINE.toString());
+        uriToOfflineSegments.put(controllerURI, offlineSegments);
+      } catch (URISyntaxException e) {
+        throw new RuntimeException("Got invalid controller uri - '" + pinotClusterSpec.getControllerURI() + "'");
+      } catch (Exception e) {
+        e.printStackTrace();
+      }
+    }
+    return uriToOfflineSegments;
+  }
+
+  /**
+   * If consistent data push is enabled, append current timestamp to existing configured segment name postfix, if
+   * configured, to make segment name unique.
+   */
+  public static void configureSegmentPostfix(SegmentGenerationJobSpec spec) {
+    SegmentNameGeneratorSpec segmentNameGeneratorSpec = spec.getSegmentNameGeneratorSpec();
+    if (consistentDataPushEnabled(spec)) {
```

Comment:
From the `SegmentGenerationJobRunner:init()` call, we already fetch table config. We should set `_consistentDataPushEnabled` there instead of fetching the table config again here.

For all cases, we should read table config only once from the controller. (once for segment generation, once for segment push)